ipd/0020/README.adoc

```diff
@@ -0,0 +1,770 @@
+----
+rpz@thunderhead:~$ pfexec ktest -o result,test,input,reason run -Ni /var/tmp/browsing.snoop mac:dummy:
+RESULT TEST                                         INPUT                                           REASON
+PASS   mac_dummy_pass_test                          --                                              --
+FAIL   mac_dummy_fail_test                          --                                              mt_dummy(5) == 0 (0x1 == 0x0) (../../common/io/mac/mac_test.c:40)
+ERROR  mac_dummy_err_test                           --                                              mt_dummy(3) != 0 (0x1 != 0x0) (../../common/io/mac/mac_test.c:47)
+SKIP   mac_dummy_skip_test                          --                                              The king stay the king.
+PASS   mac_dummy_input_test                         /var/tmp/browsing.snoop                         --
+----
+
+However, this can get a bit unwieldy, and asking ktest to print in
+parsable mode might help.
+
+.ktest run failure reason parsable mode
+----
+rpz@thunderhead:~$ pfexec ktest -po result,test,input,reason run -Ni /var/tmp/browsing.snoop mac:dummy:
+PASS:mac_dummy_pass_test::
+FAIL:mac_dummy_fail_test::mt_dummy(5) == 0 (0x1 == 0x0) (../../common/io/mac/mac_test.c\:40)
+ERROR:mac_dummy_err_test::mt_dummy(3) != 0 (0x1 != 0x0) (../../common/io/mac/mac_test.c\:47)
+SKIP:mac_dummy_skip_test::The king stay the king.
+PASS:mac_dummy_input_test:/var/tmp/browsing.snoop:
+----
+
+==== `ktest run` and `usr/src/test`
+
+The `run` command provides some great interactive support, but its
+real use is going to come from integration with `usr/src/test`. That
+is, we will use test scripts in `usr/src/test` to drive the ktest test
+modules. This is where the `-f runfile` option comes.
+
+The ktest runfile is very similar to the `run` command, except that
+the triples are specified in a file, and each triple, partially or
+fully-qualified, may be paired with its own input file.
+
+.runfile example
+----
+mac:
+mac:checksum:mac_sw_cksum_ipv4_snoop_test /var/tmp/browsing.snoop
+stream:
+----
+
+Given this file we can then run the following.
+
+----
+$ pfexec ktest run -f /var/tmp/example.run
+----
+
+
+== Related Work
+
+There are several components in illumos already that facilitate some
+of what ktest is proposing, but they are either more narrow in scope
+or lack the ability to test the kernel in-full like ktest can.
+
+=== usr/src/test
+
+This is the framework for userland testing. It provides scaffolding
+for describing, organizing, runnings, and reporting on tests. This is
+used fairly heavily by some systems to test both userland and kernel
+components. Though the later testing is of course indirect, by way of
+userland APIs, system calls, and ioctls. This framework is
+complimentary to ktest. I envision us adding tests to various sub
+directories in here where the test defines a ktest runfile for that
+specific subsystem and then executes it.
+
+=== libfakekernel
+
```

Comment:
No where did I say that libfakekernel is a bad idea. I think it's a great idea, but I also think it's objective to say that it's much more work to properly emulate the entire kernel than to just run the tests in the kernel itself, no?

As for the mdb comment, I am not implying source level debugging is bad or not useful, or even that you or anyone else should have implemented it in mdb. I feel like you are bringing some previous bad blood that I'm not privy to into this comment. My comment stems from the fact that I myself have asked Robert about mdb source-level debugging and he made it sound like something he's already hooked up before but needed time to polish and so on. So my parenthetical was more meant as "rather than emulate the entire kernel to get source-level debugging how about we add source-level debugging to mdb!" Which, by the way, I would love! I will expand on that mdb parenthetical to make it more clear what I was saying.

Reply:
I went ahead and rewrote this entire section. Apologies if my initial wording came off as dismissive or judgmental of libfakekernel. You've obviously spent a lot of time on it and it's proven very useful for smb development, and that's great. I also think source-level debugging is a fine thing to have. I hope this new section comes off better.

Reply:
Thanks.  That's better.  If it were up to me I'd just just said less about other approaches, but this is now at least neutral about other methods.  And sorry, I was "in a mood" this morning :)